tests/lax_test.py

```diff
@@ -1748,6 +1748,36 @@ def testReduce(self, op, init_val, shape, dtype, dims):
     args_maker = lambda: [rng(shape, dtype)]
     self._CompileAndCheck(fun, args_maker)
 
+  @parameterized.named_parameters(jtu.cases_from_list(
+      {"testcase_name": "_op={}_shape={}_reducedims={}_initval={}"
+       .format(op.__name__, shape, dims, init_val),
+       "op": op, "init_val": init_val, "shape": shape, "dims": dims}
+      for init_val, op in [
+        (True, lax.bitwise_and),
+        (False, lax.bitwise_and),
+        (True, lax.bitwise_or),
+        (False, lax.bitwise_or),
+        (True, lax.bitwise_xor),
+        (False, lax.bitwise_xor),
+      ]
+      for shape, dims in [
+          [(3, 4, 5), (0,)], [(3, 4, 5), (1, 2)],
+          [(3, 4, 5), (0, 2)], [(3, 4, 5), (0, 1, 2)]
+      ]))
+  def testReduceBoolean(self, op, init_val, shape, dims):
+    dtype = np.bool_
+    rng = jtu.rand_bool(self.rng())
+    init_val = np.asarray(init_val, dtype=dtype)
+    fun = lambda operand, init_val: lax.reduce(operand, init_val, op, dims)
+    args_maker = lambda: [rng(shape, dtype), init_val]
+    self._CompileAndCheck(fun, args_maker)
+
+    # we separately test the version that uses a concrete init_val because it
+    # can hit different code paths
+    fun = lambda operand: lax.reduce(operand, init_val, op, dims)
+    args_maker = lambda: [rng(shape, dtype)]
+    self._CompileAndCheck(fun, args_maker)
```

Comment:
A suggestion here: you could use `jax.make_jaxpr` and assert that the operation is implemented via the expected primitive, e.g something like
```python
prim = lax.reduce_or_p
jaxpr = jax.make_jaxpr(fun)(rng(shape, dtype))
self.assertEqual(jaxpr.eqns[0].primitive, prim)
```
where you could add the expected reduction primitive to the test parameterization.

Reply:
Thanks for the suggestion. I did that, but the primitive is "reduce" every time. However, printing the jaxpr, it looks like the correct things are being done:

```
>>> jaxpr.eqns[0]
>>> a:bool[] = reduce[
  computation=<function bitwise_xor at 0x134349790>
  consts=()
  dimensions=(0,)
  jaxpr={ lambda ; b:bool[] c:bool[]. let d:bool[] = xor b c in (d,) }
] e f
```
The value is just kept in another entry than `primitive`. Going to look and see which one it is (debugger tells me that it might be `computation`).

Reply:
It is `jaxpr.params["computation"].__name__":

```
>>> jaxpr.params["computation"].__name__
>>> 'bitwise_xor'
```

Reply:
If the bound primitive is reduce and not reduce_xor, then this test is not actually exercising the new code. We should figure out why.

Reply:
Are you sure? Because I am getting this also:

```python
>>> shape, dtype = (3, 4, 5), np.int32
>>> fun = lambda operand, init_val: lax.reduce(operand, init_val, lax.add, (0,))
>>> jax.make_jaxpr(fun)(rng(shape, dtype), 0)
{ lambda ; a:i32[3,4,5] b:i32[]. let
    c:i32[4,5] = reduce[
      computation=<function add at 0x127dfb940>
      consts=()
      dimensions=(0,)
      jaxpr={ lambda ; d:i32[] e:i32[]. let f:i32[] = add d e in (f,) }
    ] a b
  in (c,) }
>>> jax.make_jaxpr(fun)(a(shape, dtype), 0).eqns[0].primitive
>>> reduce
```

Maybe I am not doing something completely correctly here, though.

Reply:
You need the init  value to be static, so it will only work for this test and not the other:
```python
import jax
import jax.numpy as jnp

def f(x):
  return jax.lax.reduce(x, 0, jax.lax.add, (0,))

jaxpr = jax.make_jaxpr(f)(jnp.arange(4))
print(jaxpr)
# { lambda ; a:i32[4]. let b:i32[] = reduce_sum[axes=(0,)] a in (b,) }

print(jaxpr.eqns[0].primitive is jax.lax.reduce_sum_p)
# True
```

Reply:
`lax.reduce` will only produce the correct results if the `init_val` and `op` form a monoid. So we can't iterate over arbitrary init values; you have to use the correct `init_val` for the given `op`.

Reply:
Thanks! But that means that we could just delete these, since they violate the assumptions `lax.reduce` anyway. Is that correct?

Reply:
When you say "delete these", what are you referring to?

Reply:
I meant "do not run tests with the wrong `init_val` for a given op" (so `False` for `reduce_and`, `True` for `reduce_{or,xor}`). The rationale is that since the monoidal assumption to `lax.reduce` was being violated by the tests inputs (specifically the wrong `init_val` for these ops), I was actually testing the wrong thing.

Do you want non-monoidal codepaths to be tested as well? Then I could add some tests for that, too (with the primitive as `reduce_p`).

Reply:
No, i don’t think we need tests for invalid inputs